router/dataplane_test.go

```diff
@@ -49,6 +51,175 @@ import (
 	"github.com/scionproto/scion/router/mock_router"
 )
 
+var metrics = router.NewMetrics()
+
+func TestReceiver(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	prepareDP := func(ctrl *gomock.Controller) *router.DataPlane {
+		ret := &router.DataPlane{Metrics: metrics}
+
+		key := []byte("testkey_xxxxxxxx")
+		local := xtest.MustParseIA("1-ff00:0:110")
+
+		mInternal := mock_router.NewMockBatchConn(ctrl)
+		mInternal.EXPECT().ReadBatch(gomock.Any()).DoAndReturn(
+			func(m underlayconn.Messages) (int, error) {
+				for i := 0; i < 10; i++ {
+					spkt, dpath := prepBaseMsg(time.Now())
+					spkt.DstIA = local
+					dpath.HopFields = []path.HopField{
+						{ConsIngress: 41, ConsEgress: 40},
+						{ConsIngress: 31, ConsEgress: 30},
+						{ConsIngress: 1, ConsEgress: 0},
+					}
+					dpath.Base.PathMeta.CurrHF = 0
+					dpath.HopFields[0].Mac = computeMAC(t, key,
+						dpath.InfoFields[0], dpath.HopFields[2])
+					spkt.Path = dpath
+					payload := bytes.Repeat([]byte("actualpayloadbytes"), i)
+					buffer := gopacket.NewSerializeBuffer()
+					err := gopacket.SerializeLayers(buffer,
+						gopacket.SerializeOptions{FixLengths: true},
+						spkt, gopacket.Payload(payload))
+					require.NoError(t, err)
+					raw := buffer.Bytes()
+					copy(m[i].Buffers[0], raw)
+					m[i].N = len(raw)
+					m[i].Addr = &net.UDPAddr{IP: net.IP{10, 0, 200, 200}}
+				}
+				return 10, nil
+			},
+		).Times(1)
+		mInternal.EXPECT().ReadBatch(gomock.Any()).Return(0, nil).AnyTimes()
+
+		_ = ret.AddInternalInterface(mInternal, net.IP{})
+		_ = ret.SetIA(local)
+		_ = ret.SetKey(key)
+		return ret
+	}
+	dp := prepareDP(ctrl)
+	ch := dp.ConfigureProcChannels(1, 100)
+	dp.InitializePacketPool(100)
+	dp.ConfigureBatchSize(64)
+	assert.Equal(t, 100, dp.CurrentPoolSize())
+	dp.SetRunning(true)
+	dp.InitMetrics()
+	defer dp.SetRunning(false)
+	go func() {
+		dp.InitReceiver(router.NetworkInterface{InterfaceId: 0, Addr: nil, Conn: dp.GetInternalInterface()})
+	}()
+	for i := 0; i < 10; i++ {
+		select {
+		case <-ch[0]:
+			// we just check that the processing routine has received the packet
+			assert.Greater(t, 100, dp.CurrentPoolSize())
```

Comment:
We don't know this from the current pool size. This is an additional check. We know it because if the test is successful we were able to read exactly 20 times from the channel.

Reply:
I check that the receiver takes the buffers from the buffer pool, which decreases its size. Since we are not running a processing routine but are manually reading from the processing routine channel, it is not expected that the buffer pool ever increases again during the test.

Reply:
I modified the comments in the unit test.